#ifndef FRUIT_DETECTION_NODE_H
#define FRUIT_DETECTION_NODE_H


//Include openCV first, otherwise it does not compile.
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/core.hpp>
#include <cv_bridge/cv_bridge.h>


#include "HalconCpp.h"
#include "HDevEngineCpp.h"
#include "window_handling.h"

#include "ros/ros.h"
#include "sensor_msgs/Image.h"
#include <fruit_detection/request_features.h>

/*°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°
              The global variables in this frame are used in the HDevEngine.     
°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°*/
HDevEngineCpp::HDevEngine       hdevengine;
HDevEngineCpp::HDevProgram      hdev_script;
HDevEngineCpp::HDevProgramCall  script_results;

bool halcon_script_initialized;

std::string fruit_detection_halcon_script_path;
std::string fruit_detection_halcon_external_procedures_path;

std::string find_pepper_features_function;

int hue_lower_threshold;
int hue_upper_threshold;

HTuple cog_x_tuple;
HTuple cog_y_tuple;
HTuple size_tuple;

double pepper_blob_cog_x;
double pepper_blob_cog_y;
double pepper_blob_size;
/*‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡*/



/*°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°
                          Global publishers and Subscribers of the ROS node   
°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°*/
ros::ServiceServer      pepper_features_output_service;
/*‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡*/





/*°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°
                              All functions within this ros node.
°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°°*/
HalconCpp::HImage rosImage2Halcon(sensor_msgs::Image &ros_image);
bool getPepperFeatures(fruit_detection::request_features::Request &req, fruit_detection::request_features::Response &res);
void intializeHalconProcedures();
bool runHalconScript();
void initializeHalconScript();
int  main(int argc, char **argv);
/*‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡‡*/
#endif
